src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestCallback.cs

```diff
@@ -0,0 +1,360 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Net.Security;
+using System.Runtime.InteropServices;
+using System.Security.Cryptography.X509Certificates;
+
+using SafeWinHttpHandle = Interop.WinHttp.SafeWinHttpHandle;
+
+namespace System.Net.Http
+{
+    /// <summary>
+    /// Static class containing the WinHttp global callback and associated routines.
+    /// </summary>
+    internal static class WinHttpRequestCallback
+    {
+        public static Interop.WinHttp.WINHTTP_STATUS_CALLBACK s_StaticCallbackDelegate =
+            new Interop.WinHttp.WINHTTP_STATUS_CALLBACK(WinHttpCallback);
+
+        public static void WinHttpCallback(
+            IntPtr handle,
+            IntPtr context,
+            uint internetStatus,
+            IntPtr statusInformation,
+            uint statusInformationLength)
+        {
+            bool invokeCallback = false;
+
+            if (Environment.HasShutdownStarted)
```

Comment:
@davidsh Actually I'm reconsidering this: if the finalizer thread just started and we're going to track WinHttp safe handles via `WINHTTP_CALLBACK_STATUS_HANDLE_CLOSING` callbacks we'll need this code to still execute, otherwise the objects will never get to ref-count 0. This applies to code in WebSocket.Client as well.